core/server/worker/src/main/java/alluxio/worker/block/management/BlockTransferExecutor.java

```diff
@@ -0,0 +1,260 @@
+   */
+  private List<List<BlockTransferInfo>> balancePartitions(List<List<BlockTransferInfo>> partitions,
+      int partitionCount) {
+    // Return as is if less than requested bucket count.
+    if (partitions.size() <= partitionCount) {
+      return partitions;
+    }
+
+    // Greedily build a balanced partitions by transfer count.
+
+    // TODO(ggezer): Capacity based partitioning possible if BlockTransferInfo contains the block
+    // size.
+    Collections.sort(partitions, Comparator.comparingInt(List::size));
+
+    // Initialize balanced partitions.
+    List<List<BlockTransferInfo>> balancedPartitions = new ArrayList<>(partitionCount);
+    for (int i = 0; i < partitionCount; i++) {
+      balancedPartitions.add(new LinkedList<>());
+    }
+    // Place partitions into balanced partitions.
+    for (List<BlockTransferInfo> partition : partitions) {
+      // Find the balanced partition with the least element size.
+      int selectedPartitionIdx = -1;
+      int selectedPartitionCount = -1;
+      for (int i = 0; i < partitionCount; i++) {
+        if (balancedPartitions.get(i).size() > selectedPartitionCount) {
+          selectedPartitionIdx = i;
+          selectedPartitionCount = balancedPartitions.get(i).size();
+        }
+      }
+      balancedPartitions.get(selectedPartitionIdx).addAll(partition);
+    }
+
+    return balancedPartitions;
+  }
+
+  /**
+   * Used to determine right partitioning key by inspecting list of transfers.
+   */
+  private TransferPartitionKey findTransferBucketKey(List<BlockTransferInfo> transferInfos) {
+    int srcAllocatedCount = 0;
+    int dstAllocatedCount = 0;
+    for (BlockTransferInfo transferInfo : transferInfos) {
+      if (transferInfo.getSrcLocation().dir() != BlockStoreLocation.ANY_DIR) {
+        srcAllocatedCount++;
+      }
+      if (transferInfo.getDstLocation().dir() != BlockStoreLocation.ANY_DIR) {
+        dstAllocatedCount++;
+      }
+    }
+
+    if (srcAllocatedCount == dstAllocatedCount) {
+      if (srcAllocatedCount == 0) {
+        return TransferPartitionKey.NONE;
+      } else {
+        // Fall-back to SRC partitioning if all are allocated.
+        return TransferPartitionKey.SRC;
+      }
+    } else if (srcAllocatedCount > dstAllocatedCount) {
+      return TransferPartitionKey.SRC;
+    } else {
+      return TransferPartitionKey.DST;
+    }
+  }
+
+  /**
+   * Used as entry point for executing a single transfer partition.
+   */
+  private void executeTransferPartition(List<BlockTransferInfo> transferInfos,
+      Consumer<Exception> exceptionHandler) {
+    for (BlockTransferInfo transferInfo : transferInfos) {
+      try {
+        if (mLoadTracker.loadDetected(transferInfo.getSrcLocation(),
+            transferInfo.getDstLocation())) {
+          LOG.warn("Skipping transfer-order: {} due to user activity.", transferInfo);
+          continue;
+        }
+
+        boolean useReservedSpace = transferInfo.isSwap();
+
+        synchronized (this) {
```

Comment:
What is this lock intended to guard against? It looks like `BlockTransferExecutor` is never shared among threads?